lib/src/main/java/com/knuddels/jtokkit/TokenEncoderLarge.java

```diff
@@ -33,46 +33,49 @@ static int calculateTokensLarge(TokenEncoder tokenEncoder, int maxTokenCount, bo
         }
 
         while (validRanks > 0) {
-            RankNode minNode = rankMap.firstEntry().getValue().firstEntry().getValue();
-            assert minNode.rank != MAX_RANK;
-
-            RankNode previousNode = minNode.prev;
-            RankNode nextNode = minNode.next;
-            RankNode nextNextNode = nextNode != null ? nextNode.next : null;
-            RankNode nextNextNextNode = nextNextNode != null ? nextNextNode.next : null;
-
-            if (previousNode != null) {
-                int newRank = tokenEncoder.encode(match, previousNode.index, nextNextNode != null ? nextNextNode.index : Integer.MAX_VALUE);
-                if ((newRank == MAX_RANK) != (previousNode.rank == MAX_RANK)) {
-                    validRanks -= (newRank == MAX_RANK) ? 1 : -1;
+            TreeMap<Integer, RankNode> minNodes = rankMap.firstEntry().getValue();
+            for (int i = 0; i < minNodes.size(); i++) {
+                RankNode minNode = minNodes.firstEntry().getValue();
```

Comment:
the minimum we've found will be the next minimum anyway, if there are multiple tokens with the same rank, so we can just use up the gathered ones